exporters/prometheus/src/push_exporter.cc

```diff
@@ -1,23 +1,123 @@
 // Copyright 2022, OpenTelemetry Authors
 // SPDX-License-Identifier: Apache-2.0
 
-#  include "opentelemetry/exporters/prometheus/push_exporter.h"
+#include "opentelemetry/exporters/prometheus/push_exporter.h"
 
 OPENTELEMETRY_BEGIN_NAMESPACE
 
-namespace exporter {
-namespace metrics {
+namespace exporter
+{
+namespace metrics
+{
+
+class PrometheusPushCollector : public ::prometheus::Collectable
+{
+public:
+  /**
+   * Default Constructor.
+   *
+   * This constructor initializes the collection for metrics to export
+   * in this class with default capacity
+   */
+  explicit PrometheusPushCollector(size_t max_collection_size = 2048)
+      : max_collection_size_(max_collection_size)
+  {
+    metrics_to_collect_.reserve(max_collection_size_);
+  }
+
+  /**
+   * Collects all metrics data from metricsToCollect collection.
+   *
+   * @return all metrics in the metricsToCollect snapshot
+   */
+  std::vector<::prometheus::MetricFamily> Collect() const override
+  {
+    if (!collection_lock_.try_lock())
+    {
+      return {};
+    }
+
+    // copy the intermediate collection, and then clear it
+    std::vector<::prometheus::MetricFamily> moved_data;
+    moved_data.swap(metrics_to_collect_);
+    metrics_to_collect_.reserve(max_collection_size_);
+
+    collection_lock_.unlock();
+
+    return moved_data;
+  }
+
+  /**
+   * This function is called by export() function and add the collection of
+   * records to the metricsToCollect collection
+   *
+   * @param records a collection of records to add to the metricsToCollect
+   * collection
+   */
+  void AddMetricData(const ::opentelemetry::sdk::metrics::ResourceMetrics &data)
+  {
+    auto translated =
+        ::opentelemetry::exporter::metrics::PrometheusExporterUtils::TranslateToPrometheus(data);
+
+    std::lock_guard<std::mutex> guard{collection_lock_};
+
+    for (auto &item : translated)
+    {
+      if (metrics_to_collect_.size() + 1 <= max_collection_size_)
+      {
+        // We can not use initializer lists here due to broken variadic capture
+        // on GCC 4.8.5
+        metrics_to_collect_.emplace_back(std::move(item));
+      }
+    }
+  }
+
+  /**
+   * Get the current collection in the collector.
+   *
+   * @return the current metricsToCollect collection
+   */
+  std::vector<::prometheus::MetricFamily> &GetCollection() { return metrics_to_collect_; }
+
+  /**
+   * Gets the maximum size of the collection.
+   *
+   * @return max collection size
+   */
+  std::size_t GetMaxCollectionSize() const noexcept { return max_collection_size_; }
```

Comment:
Nit - we should guard GetMaxCollectionSize ad GetCollection too ?

Reply:
Maybe we can use atomic here. Can I  change it in another PR?

Reply:
Fixed